_user/rotrectangle/main.c

```diff
@@ -0,0 +1,273 @@
+{
+	double dx, dy, angleRad, xcDouble, ycDouble, aDouble, bDouble;
+	unsigned int x0, y0, x, y;		/* (x,y) is start point for next rectangle lines */
+
+	xcDouble = (double)self->xc;
+	ycDouble = (double)self->yc;
+	aDouble = (double)self->a;
+	bDouble = (double)self->b;
+
+	angleRad = ((double)(angle)*3.141592)/180.0;
+	x0 = (unsigned int)(xcDouble - ( (sin(M_PI/2-angleRad-atan(bDouble/aDouble)))*sqrt(pow(aDouble, 2.0)+pow(bDouble, 2.0)) )/2.0);
+	y0 = (unsigned int)(ycDouble - ( (cos(M_PI/2-angleRad-atan(bDouble/aDouble)))*sqrt(pow(aDouble, 2.0)+pow(bDouble, 2.0)) )/2.0);
+
+	x = x0;
+	y = y0;
+	dx = (self->a)*cos(angleRad); 
+	dy = (self->a)*sin(angleRad);
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	x = x0+dx;
+	y = y0+dy;
+	dx = (self->b)*cos(angleRad+(0.5*3.141592));
+	dy = (self->b)*sin(angleRad+(0.5*3.141592));
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	x = x0;
+	y = y0;
+	dx = (self->b)*cos(angleRad+(0.5*3.141592)); 
+	dy = (self->b)*sin(angleRad+(0.5*3.141592));
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	x = x0+dx;
+	y = y0+dy;
+	dx = (self->a)*cos(angleRad); 
+	dy = (self->a)*sin(angleRad);
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	return 0;
+}
+
+static int rotrectangle_save(rotrectangle_t *self)
+{
+	self->toRemove.xc = self->rec.xc;
+	self->toRemove.yc = self->rec.yc;
+	self->toRemove.a = self->rec.a;
+	self->toRemove.b = self->rec.b;
+	self->toRemove.color = 0;
+
+	return 0;
+}
+
+static int rotrectangle_rotating(rotrectangle_t *self, graph_t *graphp)
+{
+	unsigned int alfa, ret;
+	ret = 0;
+	while (self->stop == 0) {
+		for (alfa=0; alfa<360; alfa+=self->rotSpeed) {
+			if (self->stop != 0)
+				break;
+			ret = rotrectangle_print(&(self->rec), graphp, alfa);
+			// graph_fill(graphp, self->rec.xc, self->rec.yc, self->rec.color, GRAPH_FILL_FLOOD, GRAPH_QUEUE_HIGH);
+			ret = rotrectangle_save(self);
+			usleep(10000);		/* min value */
+			ret = rotrectangle_print(&(self->toRemove), graphp, alfa);
+			// graph_fill(graphp, self->rec.xc, self->rec.yc, 0, GRAPH_FILL_FLOOD, GRAPH_QUEUE_HIGH);
+		}
+	}
+	
+	return ret;
+}
+
+static int rotrectangle_startpanel(rotrectangle_t *self)
+{
+	printf("*** Rotating rectangle program ***\n");
+	printf("  User interface:\n");
+	printf("    )up arrow - move up\n");
+	printf("    )down arrow - move down\n");
+	printf("    )right arrow - move right\n");
+	printf("    )left arrow - move left\n");
+	printf("    )'+' - increase rotation speed\n");
+	printf("    )'-' - decrease rotation speed\n");
+	printf("    )'r' - turn red\n");
+	printf("    )'g' - turn green\n");
+	printf("    )'b' - turn blue\n");
+	printf("    )'q' - exit\n");
+	printf("  Press any key to start: \n");
+	getchar();
+
+	return 0;
+}
+
+int main(void)
+{
+	int ret;
+	char stack[1024] __attribute__ ((aligned(8)));
```

Comment:
It's better to put big memory objects in .bss instead of stack.

Reply:
This is general remark. It is even better to allocate thread stack using `malloc()`.